modules/fx/arrow-fx-reaktive/src/main/kotlin/arrow/fx/reaktive/extensions/Common.kt

```diff
@@ -0,0 +1,124 @@
+package arrow.fx.reaktive.extensions
+
+import arrow.fx.RacePair
+import arrow.fx.RaceTriple
+import arrow.fx.reaktive.connectAndGetDisposable
+import arrow.fx.typeclasses.Fiber
+import com.badoo.reaktive.coroutinesinterop.asScheduler
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.maybe.Maybe
+import com.badoo.reaktive.maybe.MaybeCallbacks
+import com.badoo.reaktive.maybe.MaybeObserver
+import com.badoo.reaktive.maybe.asObservable
+import com.badoo.reaktive.maybe.doOnBeforeFinally
+import com.badoo.reaktive.maybe.filter
+import com.badoo.reaktive.maybe.map
+import com.badoo.reaktive.maybe.maybe
+import com.badoo.reaktive.maybe.merge
+import com.badoo.reaktive.maybe.subscribeOn
+import com.badoo.reaktive.observable.ConnectableObservable
+import com.badoo.reaktive.observable.firstOrComplete
+import com.badoo.reaktive.observable.replay
+import kotlin.coroutines.CoroutineContext
+
+internal fun <F, A, B> CoroutineContext.racePair(fa: Maybe<A>, fb: Maybe<B>, fiberFactory: FiberFactory<F>): Maybe<RacePair<F, A, B>> =
+  asScheduler().let { scheduler ->
+    maybe { emitter ->
+      val connectableA: ConnectableObservable<A> = fa.subscribeOn(scheduler).asObservable().replay(bufferSize = 1)
+      val connectableB: ConnectableObservable<B> = fb.subscribeOn(scheduler).asObservable().replay(bufferSize = 1)
+      val disposableA = connectableA.connectAndGetDisposable()
+      val disposableB = connectableB.connectAndGetDisposable()
+      val maybeA = connectableA.firstOrComplete().filter { !disposableA.isDisposed }
+      val maybeB = connectableB.firstOrComplete().filter { !disposableB.isDisposed }
+      var isResultEmitted = false
+
+      merge(
+        maybeA.map { RacePairResult.First(it) },
+        maybeB.map { RacePairResult.Second(it) }
+      )
+        .firstOrComplete()
+        .map {
+          isResultEmitted = true
+          // Race condition here: the chain can be disposed before the fiber reaches its listener
+          when (it) {
```

Comment:
So here we have race condition, a few tests are flaky. Same problem with `arrow-fx-rx2`. But because of a different implementation of `Scheduler` - `CoroutineContext` - `Scheduler` interop, its tests are not flaky.
I think this is a conceptual problem for `Fiber` interop. When a `Fiber` is sent to the downstream it might not reach its listener and will not be cancelled.

Reply:
So the race-condition is that the `emitter` gets cancelled before the `Fiber` reaches the downstream listeners?
Do we have a stable reproduction scenario?

If we look at `IO#racePair` we can see that is does the following.  https://github.com/arrow-kt/arrow/blob/master/modules/fx/arrow-fx/src/main/kotlin/arrow/fx/IORace.kt#L410

- It creates an `active` flag similar to `isResultEmitted` that tracks when the result is send downstream.

- Creates 2 promises & 2 new `IOConnection`s to decouple from the existing connection, which is like `Disposable` so we decouple the 2 tasks from the original stream. In the RxJava code this translated to 2 `BehaviorSubject`s and calling `subscribe` twice to create 2 new streams. I see you solved this with `ConnectableObservable`.

- We hook `disposableA` (connA) & `disposableB` (connB) up to the original `Disposable`, using `push`. Pushing the two `IOConnection` onto the original makes sure that when the streams gets cancelled the two decoupled streams also get cancelled, the decoupling is done so the decoupled streams cannot cancel the original stream.

=> I suspect here the race condition is happening? Since this operation in `IO` makes sure that when the `emitter` is cancelled both the fibers get cancelled. Cancel is idempotent so calling it for an already cancelled `Fiber` doesn't matter. Or is that's what's happening inside line 57?

We use `emitter.setCancellable { dda.dispose(); ddb.dispose() }` for RxJava.

The difference that I know of is that we can link the cancelation for `IO` *before* we run the effect, which seems the opposite in the Rx design. So maybe we just need additional synchronization on top of it that tracks the cancellation between all running operations through `doOnDispose`?